src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.BraceCompletion
+{
+    internal abstract class AbstractBraceCompletionService : IBraceCompletionService
+    {
+        protected abstract char OpeningBrace { get; }
+
+        protected abstract char ClosingBrace { get; }
+
+        /// <summary>
+        /// Returns if the token is a valid opening token kind for this brace completion service.
+        /// </summary>
+        protected abstract bool IsValidOpeningBraceToken(SyntaxToken token);
+
+        /// <summary>
+        /// Returns if the token is a valid closing token kind for this brace completion service.
+        /// </summary>
+        protected abstract bool IsValidClosingBraceToken(SyntaxToken token);
+
+        public abstract Task<bool> AllowOverTypeAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken);
```

Comment:
definitely doc.

Reply:
I had inheritdoc on this (since it's coming from the interface), but removed in response to https://github.com/dotnet/roslyn/pull/48468#discussion_r526236784

I'll leave this as-is since there are docs on the interface method already.